src/client/Client.cc

```diff
@@ -12636,8 +12636,20 @@ int Client::_rename(Inode *fromdir, const char *fromname, Inode *todir, const ch
       fromdir->quota.is_enable() ? fromdir : get_quota_root(fromdir, perm);
     Inode *todir_root =
       todir->quota.is_enable() ? todir : get_quota_root(todir, perm);
-    if (fromdir_root != todir_root) {
-      return -EXDEV;
+    if (todir_root->quota.is_enable() && fromdir_root != todir_root) {
+      if ((todir_root->quota.max_bytes && 
+	    (fromdir->rstat.rbytes + todir_root->rstat.rbytes) 
+	      >= todir_root->quota.max_bytes) || 
+          (todir_root->quota.max_files && 
+	    (fromdir->rstat.rsize() + todir_root->rstat.rsize())
+	      >= todir_root->quota.max_files)) {
```

Comment:
Also, from a code hygiene standpoint, this if statement (two actually!) is too complex. Can you break it into several bool variables and then do the test.

Reply:
will do